core/wasmWasi/src/files/FileSystemWasm.kt

```diff
@@ -5,36 +5,539 @@
+            }.toBitset()
 
-    override fun delete(path: Path, mustExist: Boolean) = unsupported()
+            val openFlags = buildList {
+                add(OpenFlags.creat)
+                if (!append) {
+                    add(OpenFlags.trunc)
+                }
+            }.toBitset()
 
-    override fun createDirectories(path: Path, mustCreate: Boolean) = unsupported()
+            val res = Errno(
+                path_open(
+                    fd = root.fd,
+                    dirflags = listOf(LookupFlags.symlink_follow).toBitset(),
+                    pathPtr = stringBuffer.address.toInt(), pathLen = stringBufferLength,
+                    oflags = openFlags,
+                    fsRightsBase = listOf(Rights.fd_write, Rights.fd_sync).toBitset(),
+                    fsRightsInheriting = 0,
+                    fdFlags = fdFlags,
+                    resultPtr = fdPtr.address.toInt()
+                )
+            )
+            if (res != Errno.success) {
+                throw IOException("Can't open $path for write: ${res.description}")
+            }
+            fdPtr.loadInt()
+        }
+        return FileSink(fd)
+    }
 
-    override fun atomicMove(source: Path, destination: Path) = unsupported()
+    override fun metadataOrNull(path: Path): FileMetadata? {
+        val root = PreOpens.getRootOrNull(path) ?: return null
+        val md = metadataOrNullInternal(root.fd, path, true) ?: return null
 
-    override fun source(path: Path): RawSource = unsupported()
+        val filetype = md.filetype
+        val isDirectory = filetype == FileType.directory
+        val filesize = if (isDirectory) -1 else md.filesize
+        return FileMetadata(
+            isRegularFile = filetype == FileType.regular_file,
+            isDirectory = isDirectory,
+            size = filesize
+        )
+    }
 
-    override fun sink(path: Path, append: Boolean): RawSink = unsupported()
+    /**
+     * Returns an absolute path to the same file or directory the [path] is pointing to.
+     * All symbolic links are solved, extra path separators and references to current (`.`) or
+     * parent (`..`) directories are removed.
+     * If the [path] is a relative path then it'll be resolved against current working directory.
+     * If there is no file or directory to which the [path] is pointing to then [FileNotFoundException] will be thrown.
+     *
+     * The behavior of this method differs from other platforms as the resolution
+     * may not fail if there is no filesystem-node (file, directory, symlink, etc.) corresponding
+     * to some interior path. This allows successfully resolving paths like `/a/b/c/../../d/e` when
+     * pre-opened directories are `/a/b/c` and `/a/d/e`.
+     *
+     * @param path the path to resolve.
+     * @return a resolved path.
+     * @throws FileNotFoundException if there is no file or directory corresponding to the specified path.
+     */
+    override fun resolve(path: Path): Path {
```

Comment:
@dkhalanskyjb (you asked, I tagged). 
Here's a hand-written canonization implementation. It's used only for Wasm-WASI.